dev/PushNotifications/GetRawNotificationEventArgs.h

```diff
@@ -0,0 +1,32 @@
+﻿#include <pch.h>
+#include "PushNotificationReceivedEventArgs.h"
+#include "externs.h"
+
+using namespace winrt;
+using namespace winrt::Windows::Foundation;
+constexpr PCWSTR pushServerArgument = L"-ReunionPushServer";
+
+namespace GetRawNotificationEventArgs
+{
+    static Windows::Foundation::IInspectable GetRawNotificationEventArgs()
+    {
+        winrt::Microsoft::Windows::PushNotifications::PushNotificationReceivedEventArgs args = nullptr;
+
+        std::wstring commandLineArgs = GetCommandLine();
+        auto found = commandLineArgs.find(pushServerArgument, 0);
+        if (found != std::wstring::npos)
+        {
+            {
+                auto lock = g_lock.lock();
+                THROW_HR_IF_NULL_MSG(E_UNEXPECTED, g_waitHandleForArgs, "PushNotificationManager::RegisterActivator has not been called.");
+            }
+
+            if (WaitForSingleObject(g_waitHandleForArgs.get(), 1000) == WAIT_OBJECT_0)
+            {
+                auto lock = g_lock.lock();
+                std::swap(args, g_activatedEventArgs);
```

Comment:
Correct. The recommended pattern for the API is to call GetRawNotificationEventArgs only once during winmain entry. So the call above would be an anti-pattern. I guess we could also change this to always return the cached raw event args but I'm not sure if that is very valuable in the larger scheme of things.

@pmpurifoy instead of swapping this, you could directly return g_activatedEventArgs. In addition, you will need a check in the beginning of the function to see if it is nullptr. If it is not, simply just return it.

Reply:
Forbidding GetRawNotificationEventArgs from being called more than once creates artificial barriers to development.

1. You cannot have a debugger extension that dumps the current raw notification event args, because the debugger extension will call GRNEA, dump the args, and then when you resume the app, it gets GRNEA and gets `null` back.
2. Apps cannot combine components from unrelated libraries that each use raw notifications for their own purposes. E.g., I buy a "high score" library that uses raw notifications to tell me when your high score has been beaten. Also I buy a "game matchmaker" library that uses raw notifications to help me find people to play head-to-head games against. I cannot use both of these libraries in the same app, even though they can coexist (since each one ignores the raw notifications it doesn't understand).